src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -0,0 +1,304 @@
+// "X says blah(args)" into "says_blah(X, args)".
+Predicate PushOntoPredicate(absl::string_view modifier,
+                            std::vector<std::string> new_args,
+                            const Predicate& predicate) {
+  std::string new_name = absl::StrCat(std::move(modifier), predicate.name());
+  // This seemingly pointless line is needed to help C++ figure out the type
+  // of predicate.args() when it is passed to MoveAppend:
+  std::vector<std::string> pred_args = std::move(predicate.args());
+  MoveAppend(new_args, std::move(pred_args));
+  Sign sign_copy = predicate.sign();
+  return Predicate(new_name, std::move(new_args), sign_copy);
+}
+
+// This function is an abbreviation for `PushPrincipal` where:
+// - a modifier is added
+// - just one new principal is added as an argument.
+// This is a common case in this translation because it is used for
+// `x says blah(args)` and `x canActAs y` and other constructions involving a
+// principal name.
+Predicate PushPrincipal(absl::string_view modifier, const Principal& principal,
+                        const Predicate& predicate) {
+  return PushOntoPredicate(modifier, {principal.name()}, predicate);
+}
+
+Predicate AttributeToDLIR(const Attribute& attribute) {
+  // If attribute is `X pred(args...)` the following predicate is
+  // `pred(X, args...)`
+  return PushPrincipal(std::string(""), attribute.principal(),
+                       attribute.predicate());
+}
+
+DLIRAssertion LoweringToDatalogPass::SpokenAttributeToDLIR(
+    const Principal& speaker, const Attribute& attribute) {
+  Predicate main_predicate = AttributeToDLIR(attribute);
+
+  // Attributes interact with "canActAs" because if "Y canActAs X"
+  // then Y also picks up X's attributes. We need to generate
+  // an additional rule to implement this behavior. If the attribute
+  // under translation is `X PredX`, the additional rule is:
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X PredX`
+  // (Where Y is a fresh variable)
+  Principal prin_y(FreshVar());
+
+  // This is `speaker says Y PredX`
+  Predicate generated_lhs = PushPrincipal("says_", prin_y, main_predicate);
+
+  Predicate y_can_act_as_x("canActAs",
+                           {prin_y.name(), attribute.principal().name()},
+                           Sign::kPositive);
+
+  Predicate speaker_says_y_can_act_as_x =
+      PushPrincipal("says_", speaker, y_can_act_as_x);
+
+  // This is `speaker says X PredX`
+  Predicate speaker_says_x_pred =
+      PushPrincipal("says_", speaker, main_predicate);
+
+  // This is the full generated rule:
+  // `speaker says Y PredX :-
+  //    speaker says Y canActAs X, speaker says X PredX`
+  return DLIRAssertion(
+      DLIRCondAssertion(generated_lhs, {std::move(speaker_says_y_can_act_as_x),
+                                        std::move(speaker_says_x_pred)}));
+}
+
+Predicate CanActAsToDLIR(const CanActAs& can_act_as) {
```

Comment:
Move this along with other anonymous functions in this file..

Reply:
done